src/channel_state.ts

```diff
@@ -495,35 +453,50 @@ export class ChannelState<
       >['messageToImmutable']
     >,
   ) {
-    let updated = false;
-    let newMessages: Immutable.ImmutableArray<ReturnType<
-      ChannelState<
-        AttachmentType,
-        ChannelType,
-        CommandType,
-        EventType,
-        MessageType,
-        ReactionType,
-        UserType
-      >['messageToImmutable']
-    >> = Immutable([]);
-
-    for (let i = 0; i < messages.length; i++) {
-      const message = messages[i];
-      const idMatch = !!message.id && !!newMessage.id && message.id === newMessage.id;
+    // for empty list just concat and return
+    if (messages.length === 0) return messages.concat(message);
+
+    // if message is newer than last item in the list just concat and return
+    if (messages[messages.length - 1].created_at.getTime() < message.created_at.getTime())
+      return messages.concat(message);
+
+    // find the closest index to push the new message
+    let left = 0;
+    let middle = 0;
+    let right = messages.length - 1;
+    while (left <= right) {
+      middle = Math.floor((right + left) / 2);
```

Comment:
In most cases wouldn't a bottom up approach be faster than a binary search given the likelihood of a message being further up is less?

Reply:
L462 takes care of that. cases like pagination could add some 100 more messages to the other end of the array which will be O(n), right now this part is O(log(n)) which is quite fast already